From c87e9b7a88f8bb6a936f728d0e0f1a878899ae40 Mon Sep 17 00:00:00 2001
From: Zhan Lou <zhan.lou@cixtech.com>
Date: Tue, 19 Nov 2024 17:48:30 +0800
Subject: [PATCH] DPTSW-13428: v4l2: adjust large buffer size

Adjust large buffer size of bitstream buffer from 256M to 128M as the
hardware virtual address space for bistream buffer could be small. So we
allocate 6 or more bitstream buffers only when its size is smaller than
or equals to 128MB.

Signed-off-by: Zhan Lou <zhan.lou@cixtech.com>
Change-Id: I20cbdc5b15ac8b066118f171d40bfbb47315b4a0

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index ea8781dbf3..70ee29a3f6 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -6086,7 +6086,7 @@ gst_v4l2_object_decide_allocation (GstV4l2Object * obj, GstQuery * query)
           FALSE);
     }
 
-  } else if (obj->info.vinfo.size <= GST_V4L2_LARGE_BUFFER) {
+  } else if (obj->info.vinfo.size <= GST_V4L2_LARGE_BUFFER (obj)) {
     /* In this case we'll have to configure two buffer pool. For our buffer
      * pool, we'll need what the driver one, and one more, so we can dequeu */
     own_min = obj->min_buffers + 1;
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index fc602601ba..6a688fb9e2 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -42,11 +42,13 @@ typedef struct _GstV4l2ObjectClassHelper GstV4l2ObjectClassHelper;
 
 #include <gstv4l2bufferpool.h>
 
-#define GST_V4L2_LARGE_BUFFER (1<<28) /* 2^28 = 256M */
+#define GST_V4L2_LARGE_BUFFER(obj) \
+    (obj->info.vinfo.finfo && GST_VIDEO_INFO_FORMAT (&obj->info.vinfo) == GST_VIDEO_FORMAT_ENCODED ? \
+    (1<<27) : (1<<28))  /* 2^28 = 256M, 2^27 = 128M */
 
 /* size of v4l2 buffer pool in streaming case, obj->info needs to be valid */
 #define GST_V4L2_MIN_BUFFERS(obj) \
-    (obj->info.vinfo.size > GST_V4L2_LARGE_BUFFER ? 1 : 6)
+    (obj->info.vinfo.size > GST_V4L2_LARGE_BUFFER (obj) ? 1 : 6)
 
 /* max frame width/height */
 #define GST_V4L2_MAX_SIZE (1<<15) /* 2^15 == 32768 */
